### PR TITLE
Cross-build (most projects) for Scala 2.12.0

### PR DESCRIPTION
**Pull request by JoshRosen**
This patch allows several of the chill modules to be built for Scala 2.12.0.

Several subprojects, such as `chill-akka` and `chill-bijection`, cannot currently be published for 2.12 because their dependencies have not been published yet. To work around this, I used the [`sbt-doge`](https://github.com/sbt/sbt-doge) plugin, which overrides `++` and `+` to respect subprojects' `crossScalaVersions` settings, allowing you to run

```
> ; + clean; + test; + publishLocal
```

to test and publish for all Scala versions.

Refs #252.

### CODE REVIEW

**Comment by JoshRosen** [comment]
```
java.lang.NoClassDefFoundError: java/lang/invoke/SerializedLambda
    at com.esotericsoftware.kryo.serializers.ClosureSerializer.<clinit>(ClosureSerializer.java:42)
    at com.twitter.chill.AllScalaRegistrar.apply(ScalaKryoInstantiator.scala:207)
    at com.twitter.chill.ScalaKryoInstantiator.newKryo(ScalaKryoInstantiator.scala:97)
    at com.twitter.chill.avro.AvroSerializerSpec$$anonfun$5.apply(AvroSerializerSpec.scala:34)
    at com.twitter.chill.avro.AvroSerializerSpec$$anonfun$5.apply(AvroSerializerSpec.scala:34)
    at com.twitter.chill.package$$anon$1.newKryo(package.scala:30)
    at com.twitter.chill.KryoPool$2.newInstance(KryoPool.java:59)
    at com.twitter.chill.KryoPool$2.newInstance(KryoPool.java:57)
    at com.twitter.chill.ResourcePool.borrow(ResourcePool.java:37)
```

It looks like I need to add logic to not register `ClosureSerializer` unless we're running on a Java 8 JRE.

File: chill-scala/src/main/scala/com/twitter/chill/ScalaKryoInstantiator.scala [hunk 1]
```diff
@@ -201,5 +202,13 @@ class AllScalaRegistrar extends IKryoRegistrar {
     val boxedUnit = scala.Unit.box(())
     k.register(boxedUnit.getClass, new SingletonSerializer(boxedUnit))
     PackageRegistrar.all()(k)
+
+    // Enable Java 8 lambda serialization only if we are running on a Java 8 JRE:
```
Comment (JoshRosen) [inline chill-scala/src/main/scala/com/twitter/chill/ScalaKryoInstantiator.scala]: Quick question: what's the best place to register this? And what happens if it already happens to have been registered by some other means?

Comment (johnynek) [inline chill-scala/src/main/scala/com/twitter/chill/ScalaKryoInstantiator.scala]: I think we should make a `Java8ClosureRegistrar` that is an `IKryoRegistrar`: https://github.com/twitter/chill/blob/develop/chill-java/src/main/java/com/twitter/chill/IKryoRegistrar.java and it should be a no-op when we are not on java8 (as you have below).

See for instance the `PackageRegistrar.all()` above.

File: chill-scala/src/test/scala/com/twitter/chill/KryoSpec.scala [hunk 2]
```diff
@@ -299,7 +299,7 @@ class KryoSpec extends WordSpec with Matchers with BaseProperties {
       toList(rt(qr)) should equal(qrlist)
     }
     "Ranges should be fixed size" in {
-      val MAX_RANGE_SIZE = 188 // what seems to be needed.
+      val MAX_RANGE_SIZE = 447 // what seems to be needed.
```
Comment (johnynek) [inline chill-scala/src/test/scala/com/twitter/chill/KryoSpec.scala]: that's a bummer.

**Comment by johnynek** [comment]
looks great. I don't want to merge pointing to a `SNAPSHOT` so whenever we can fix that, we should merge.

File: chill-scala/src/test/scala/com/twitter/chill/KryoSpec.scala (hunk 2 above)
Comment (JoshRosen) [inline chill-scala/src/test/scala/com/twitter/chill/KryoSpec.scala]: :frowning:

We could make this constant a function of Scala version if you'd like to keep using the old value for the old build matrix entries.

File: chill-scala/src/main/scala/com/twitter/chill/ScalaKryoInstantiator.scala (hunk 1 above)
Comment (JoshRosen) [inline chill-scala/src/main/scala/com/twitter/chill/ScalaKryoInstantiator.scala]: Good idea; will do.

File: chill-scala/src/main/scala/com/twitter/chill/ScalaKryoInstantiator.scala [hunk 3]
```diff
@@ -198,8 +199,16 @@ class AllScalaRegistrar extends IKryoRegistrar {
       .forSubclass[scala.Enumeration#Value](new EnumerationSerializer)
 
     // use the singleton serializer for boxed Unit
-    val boxedUnit = scala.Unit.box(())
+    val boxedUnit = scala.runtime.BoxedUnit.UNIT
```
Comment (lrytz) [inline chill-scala/src/main/scala/com/twitter/chill/ScalaKryoInstantiator.scala]: this won't be necessary in M5, fixed in https://github.com/scala/scala/pull/5100

**Comment by johnynek** [comment]
do you have any interest in this still?

The PR was red, and I didn't look into it much till now. It looks like it may have been a transient error with Travis. I restarted it.

**Comment by johnynek** [comment]
if you can merge with develop and bump to a non-snapshot version of kryo 3.x.y I'll merge this.

**Comment by jodersky** [comment]
I think there is a chain of issues blocking progress on this:

- Scala 2.12 requires at least Java 1.8
- Full support for Java 1.8 is introduced in Kryo 3.1
- Kryo 3.1 was never released and moved straight to 4.0.0

#258 would solve the kryo issue and enable unrolling this blocker chain

**Comment by jodersky** [comment]
I think kryo 3.1 is required as it solves an issue with registering closure serializers (EsotericSoftware/kryo#299). Unfortunately 3.1 was never released, and the project went straight ahead to 4.0.0.
#258 would fix the issue and unblock this PR

**Comment by magro** [comment]
@johnynek Out of interest, is there anything holding you back to switch to kryo 4.0.0?

**Comment by johnynek** [comment]
So, sadly, many people against our recommendation have used kryo
serialization for long term storage. This creates a large anxiety for such
users to upgrade versions due to fear they are going to subtly break their
data pipelines.

I'm happy to help merge and release chill for later Kryos (basically we
increment a version each time we do that). It's a pain to have several
concurrent versions running though.

So, I'm happy to make a new version of chill with the latest kryo.

**Comment by johnynek** [comment]
Do you want to bump this up to 2.12.0?

**Comment by magro** [comment]
@johnynek Ok, I fully understand this. Thanks for taking the route to kryo 4!

**Comment by JoshRosen** [comment]
I just updated this to build against 2.12.0 final. Once #258 is merged I can update this to use Kryo 4. @johnynek, I'm going to click the "Allow edits from maintainers" button in case you want to update and merge this after #258 goes in. Otherwise, I can do it myself.

**Comment by johnynek** [comment]
okay, merged #258. if you merge develop, I'll merge this.

File: chill-scala/src/main/scala/com/twitter/chill/ScalaKryoInstantiator.scala (hunk 1 above)
Comment (JoshRosen) [reply chill-scala/src/main/scala/com/twitter/chill/ScalaKryoInstantiator.scala]: Whoops, forgot about this comment until now. I've created a `Java8ClosureRegistrar` to abstract away this code.

**Comment by JoshRosen** [comment]
This is now passing tests and I think should be ready for a final review. Let me know if you'd like me to make any changes.

**Comment by sritchie-stripe** [comment]
LGTM, but let's wait on @johnynek for the final shipit.

Once scala 2.12 ships do we have any plans to remove support for 2.10, or are we committed to three minor versions?

**Comment by johnynek** [comment]
Okay. This looks good. Let's merge this and this can be the 0.9.x branch for Kryo 4.x

**Pull request merged**